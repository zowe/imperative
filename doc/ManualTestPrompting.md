# Imperative CLI - Manual testing of prompting

This document provides commands to provide manual testing of command prompting.
To perform automated tests on prompting, you can use the `echo` shell command.
For example: `echo SYS1.PROCLIB | zowe files list ds PROMPT*`

NOTE: The default string to indicate prompt for the value is 'prompt*'.  This is case-insensitive.

## Simple

### Command:  

zowe files list ds PrOMpT*

### Prompt:

"dataSetName" Description: The name or pattern of the data set that you want to list\
Please enter "dataSetName":

### Response:
```
XXXXXXX\
XXXXXXX.A1543723.A035409\
XXXXXXX.A1543723.A035409.DATA\
XXXXXXX.A1543723.A035409.INDEX\
XXXXXXX.CLIST\
XXXXXXX.CNTL\
XXXXXXX.ISPF.ISPPROF\
XXXXXXX.RECEIVE.MAIL\
XXXXXXX.SYSUDUMP\
XXXXXXX.TEST.CNTL\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657.DATA\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657.INDEX\
```
### -----------------------------------------------------------------------------------------------------------------------------------------------

### Command:  

zowe files list ds prompt* --max prompt*

### Prompt:

"dataSetName" Description: The name or pattern of the data set that you want to list\
Please enter "dataSetName":\
"max-length" Description: The option --max-length specifies the maximum number of items to return. Skip this parameter to return all items. If you specify an incorrect value, the parameter returns up to 1000 items.\
Please enter "max-length":

### Response (max-length = 5):
```
XXXXXXX\
XXXXXXX.A1543723.A035409\
XXXXXXX.A1543723.A035409.DATA\
XXXXXXX.A1543723.A035409.INDEX\
XXXXXXX.CLIST\
```
### -----------------------------------------------------------------------------------------------------------------------------------------------

### Command:  

zowe files list ds prompt* --user prompt* --password prompt*

### Prompt:

"dataSetName" Description: The name or pattern of the data set that you want to list\
Please enter "dataSetName":\
"user" Description: Mainframe (z/OSMF) user name, which can be the same as your TSO login.\
Please enter "user":\
"password" Description: Mainframe (z/OSMF) password, which can be the same as your TSO password.\
Please enter "password":

### Response:
```
XXXXXXX\
XXXXXXX.A1543723.A035409\
XXXXXXX.A1543723.A035409.DATA\
XXXXXXX.A1543723.A035409.INDEX\
XXXXXXX.CLIST\
XXXXXXX.CNTL\
XXXXXXX.ISPF.ISPPROF\
XXXXXXX.RECEIVE.MAIL\
XXXXXXX.SYSUDUMP\
XXXXXXX.TEST.CNTL\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657.DATA\
XXXXXXX.ZOSFILE.VSAM.A1548874.A540657.INDEX\
```
### -----------------------------------------------------------------------------------------------------------------------------------------------

### Command:  

zowe jobs list jobs --owner prompt*

### Prompt:

"owner" Description: Specify the owner of the jobs you want to list. The owner is the individual/user who submitted the job OR the user ID assigned to the job. The command does not prevalidate the owner. You can specify a wildcard according to the z/OSMF Jobs REST
endpoint documentation, which is usually in the form "USER*".\
Please enter "owner":

### Response:
```
TSU29166   XXXXX03 ACTIVE
JOB30155   XXXXX0G INPUT
JOB29783   XXXXX0G INPUT
JOB29478   XXXXX0G INPUT
```
### -----------------------------------------------------------------------------------------------------------------------------------------------

### Command:  

zowe ims query program prompt*

### Prompt:

"names" Description: Specifies the names of the programs to query.\
Please enter "names":

### Response:
```
dopt cc bmpt dfnt    gpsb fp tmim                 rgnt pgm      schd   tls lrsdnt mbr  tmcr\
N    0  Y    MODBLKS N    N  2019.080 05:48:43.58 BMP  DFSSAM01 SERIAL N   N      IMJJ 2019.032 05:12:03.54\
N    0  N    MODBLKS N    N  2019.080 05:48:43.58 MPP  DFSSAM05 SERIAL N   N      IMJJ 2019.032 05:12:03.54\
```

### -----------------------------------------------------------------------------------------------------------------------------------------------

Over-riding the default prompt string.

Define the environmental variable, ZOWE_PROMPT_PHRASE, and set it's value to 'pmpt#'.

### Command:  

zowe files list ds pmpt# --max pmpt#

### Prompt:

"dataSetName" Description: The name or pattern of the data set that you want to list
Please enter "dataSetName":
"max-length" Description: The option --max-length specifies the maximum number of items to return. Skip this parameter to return all items. If you specify an incorrect value, the parameter returns up to 1000 items.
Please enter "max-length":

### Response:
```
XXXXXXX
XXXXXXX.A1543723.A035409
XXXXXXX.A1543723.A035409.DATA
XXXXXXX.A1543723.A035409.INDEX
XXXXXXX.CLIST

```