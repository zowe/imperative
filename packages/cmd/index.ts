/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/doc/profiles/definition/ICommandProfileSchema";
export * from "./src/doc/profiles/definition/ICommandProfileProperty";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";
export * from "./src/doc/handler/ICommandHandler";
export * from "./src/doc/handler/ICommandHandlerResponseChecker";
export * from "./src/doc/handler/ICommandHandlerResponseValidator";
export * from "./src/doc/handler/IHandlerParameters";
export * from "./src/doc/handler/IChainedHandlerEntry";
export * from "./src/doc/handler/IChainedHandlerArgumentMapping";

export * from "./src/doc/option/ICommandOptionAllowableValues";
export * from "./src/doc/option/ICommandOptionDefinition";
export * from "./src/doc/option/ICommandOptionValueImplications";
export * from "./src/doc/option/ICommandPositionalDefinition";

export * from "./src/doc/response/response/ICommandResponse";
export * from "./src/doc/response/parms/ICommandResponseParms";
export * from "./src/doc/response/response/ICommandValidatorError";
export * from "./src/doc/response/response/ICommandValidatorResponse";

export * from "./src/doc/ICommandDefinition";
export * from "./src/doc/ICommandDefinitionPassOn";
export * from "./src/doc/ICommandExampleDefinition";
export * from "./src/doc/IPartialCommandDefinition";

export * from "./src/doc/args/ICommandArguments";

export * from "./src/handlers/FailedCommandHandler";

export * from "./src/help/abstract/AbstractHelpGenerator";
export * from "./src/help/abstract/AbstractHelpGeneratorFactory";

export * from "./src/help/doc/IHelpGeneratorFactory";
export * from "./src/help/doc/IHelpGeneratorFactoryParms";
export * from "./src/help/doc/IHelpGeneratorParms";
export * from "./src/help/DefaultHelpGenerator";
export * from "./src/help/HelpConstants";
export * from "./src/help/HelpGeneratorFactory";
export * from "./src/help/doc/IHelpGenerator";
export * from "./src/help/WebHelpGenerator";
export * from "./src/help/WebHelpManager";

export * from "./src/doc/profiles/definition/ICommandProfile";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";

export * from "./src/profiles/CliProfileManager";

export * from "./src/syntax/SyntaxValidator";

export * from "./src/utils/CommandUtils";
export * from "./src/utils/SharedOptions";

export * from "./src/yargs/doc/IYargsParms";
export * from "./src/yargs/doc/IYargsResponse";

export * from "./src/yargs/AbstractCommandYargs";
export * from "./src/yargs/CommandYargs";
export * from "./src/yargs/GroupCommandYargs";
export * from "./src/yargs/YargsConfigurer";
export * from "./src/yargs/YargsDefiner";

export * from "./src/CommandPreparer";
export * from "./src/CommandProcessor";
export * from "./src/response/CommandResponse";

export * from "./src/profiles/CommandProfiles";

export * from "./src/response/CommandResponse";
export * from "./src/response/HandlerResponse";

export * from "./src/doc/response/api/handler/IHandlerResponseApi";
export * from "./src/doc/response/api/handler/IHandlerProgressApi";
export * from "./src/doc/response/api/handler/IHandlerResponseConsoleApi";
export * from "./src/doc/response/api/handler/IHandlerResponseDataApi";
export * from "./src/doc/response/api/handler/IHandlerFormatOutputApi";

export * from "./src/doc/response/response/ICommandOutputFormat";
