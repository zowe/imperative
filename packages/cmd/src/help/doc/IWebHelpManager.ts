import { IHandlerResponseApi } from "../../doc/response/api/handler/IHandlerResponseApi";

/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Web help manager API that handles launching of web help and generating it if necessary.
 * @export
 * @interface IWebHelpManager
 */
export interface IWebHelpManager {
    /**
     * Launches root help page in browser.
     * @memberof IWebHelpManager
     */
    openRootHelp(cmdResponse: IHandlerResponseApi): void;

    /**
     * Launches help page for specific group/command in browser.
     * @param {string} inContext - Name of the page to load, passed as a URL param
     * @memberof IWebHelpManager
     */
    openHelp(inContext: string, cmdResponse: IHandlerResponseApi): void;
}
