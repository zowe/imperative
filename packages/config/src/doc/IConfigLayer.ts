/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IConfig } from "./IConfig";

export interface IConfigLayerInfo {
    global: boolean;
    user: boolean;
}

export interface IConfigLayerLoc {
    global?: boolean;
    user?: boolean;
    path: string;
}

export interface IConfigLayer extends IConfigLayerInfo {
    path: string;
    exists: boolean;
    properties: IConfig;
}
