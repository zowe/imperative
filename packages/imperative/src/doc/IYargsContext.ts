/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as net from "net";


/**
 * Allow for passing our own "context" / user data through yargs
 * @export
 * @interface IYargsContext
 */
export interface IYargsContext {

    /**
     * Stream to write response to
     * @type {net.Socket}
     * @memberof IYargsContext
     */
    stream?: net.Socket;

    /**
     * Current working directory from socket client
     * @type {string}
     * @memberof IYargsContext
     */
    cwd?: string;
}