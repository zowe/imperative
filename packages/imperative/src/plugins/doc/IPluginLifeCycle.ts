/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IHandlerResponseApi } from "../../../../cmd/src/doc/response/api/handler/IHandlerResponseApi";

/**
 * An optional module of a plugin adheres to this interface to
 * perform actions during the installation lifecycle of a plugin.
 */
export interface IPluginLifeCycle {
    /**
     * This function will be called after a plugin is installed.
     * A plugin can use this opportunity to perform a sanity test or to
     * perform some additional setup which is specific to that plugin.
     * A plugin which provides a specialized credential manager should use
     * this opportunity to insert itself as an override of the standard
     * credential manager that is delivered with Zowe CLI.
     *
     * @throws
     *      A ImperativeError containing a message describing any error
     *      that occurred while performing post-install actions.
     */
    postInstall(): void;

    /**
     * This function will be called before a plugin is uninstalled.
     * This lifecycle hook is intended to replace the capability that used to
     * be performed by the NPM pre-uninstall action before NPM removed that
     * capability in NPM version 7.
     * See https://docs.npmjs.com/cli/v9/using-npm/scripts#a-note-on-a-lack-of-npm-uninstall-scripts
     *
     * A plugin can use this opportunity to revert any specialized setup that was
     * established during the lifetime of the plugin. A plugin which provides a
     * specialized credential manager should use this opportunity to remove itself as
     * an override of the standard credential manager that is delivered with Zowe CLI.
     *
     * @throws
     *      A ImperativeError containing a message describing any error
     *      that occurred while performing post-install actions.
     *
     * After an error occurs, the CLI will NOT perform the un-installation of the plugin.
     */
    preUnstall(): void;
}
