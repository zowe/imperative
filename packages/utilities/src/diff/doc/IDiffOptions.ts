/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

type outputFormat = "html" | "unifiedstring" | "terminal";

/**
 * interface for diff options
 */
export interface IDiffOptions {

    /**
     * Output format of differences between two, to be returned
     * @type {outputFormat}
     * @memberOf IDiffOptions
     */
    outputFormat: outputFormat,

    /**
     * Number of context line arguments
     * @type {number}
     * @memberOf IDiffOptions
     */
    contextLinesArg?: number,

    /**
     * Optional name for string1
     * @type {string}
     * @memberOf IDiffOptions
     */
    name1?: string,

    /**
     * Optional name for string2
     * @type {string}
     * @memberOf IDiffOptions
     */
    name2?: string,
}