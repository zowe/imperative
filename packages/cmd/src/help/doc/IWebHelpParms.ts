/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandDefinition } from "../../doc/ICommandDefinition";

/**
 * This is the set of parameters used by WebHelp modules.
 * They are supplied by the imperative module.
 */
export interface IWebHelpParms {
    /**
     * Contents of the package.json of our calling CLI.
     */
    callerPackageJson: any;

    /**
     * The path to the user's CLI home directory for the CLI's runtime files.
     */
    cliHome: string;

    /**
     * The home directory for your CLI's configuration, logging,
     * extensions, etc.
     *     e.g.  "~"/.myapp"
     * Defaults to ~/.yourcliname
     *
     * Todo: Determine why WebHelp uses both cliHome and defaultHome.
     */
    defaultHome: string;

    /**
     * A reference to our CLI's full command tree.
     */
    fullCommandTree: ICommandDefinition;

    /**
     * The display name for your CLI, used in messages.
     */
    productDisplayName: string;

    /**
     * The CLI's top-level command name. It is the first property name under
     * the 'bin' property in the CLI's package.json file.
     */
    rootCommandName: string;

    /**
     * The description that will be displayed if the user
     * gets help on the CLI's root command.
     */
    rootCommandDescription: string;

    /**
     * Path to an image of a logo for a CLI.
     * It will be displayed at the top of web help pages.
     */
    webHelpLogoImgPath: string;

    /**
     * Path to a custom CSS file for web help.
     * It will replace the main.css file that controls the style of the page.
     */
    webHelpCustomCssPath: string;
}
