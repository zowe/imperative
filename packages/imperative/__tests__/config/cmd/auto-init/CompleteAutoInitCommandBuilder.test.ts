/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { CompleteAutoInitCommandBuilder } from "../../../../src/config/cmd/auto-init/builders/CompleteAutoInitCommandBuilder";
import { Logger } from "../../../../../logger";
import { ICommandDefinition } from "../../../../../cmd";
import { fakeAutoInitConfig } from "./__data__/SampleAutoInitConfig";
import { IImperativeAutoInitCommandConfig } from "../../../../src/doc/IImperativeAutoInitCommandConfig";

const autoInitCommandConfig: IImperativeAutoInitCommandConfig = {
    autoInitConfig: fakeAutoInitConfig
};

describe("CompleteAutoInitCommandBuilder", () => {
    it("should create complete auth group given only an auto init config object", () => {
        const cmdDef: ICommandDefinition = CompleteAutoInitCommandBuilder.getAutoInitCommand(autoInitCommandConfig.autoInitConfig,
            Logger.getImperativeLogger());
        expect(cmdDef).toMatchSnapshot();
    });
});
