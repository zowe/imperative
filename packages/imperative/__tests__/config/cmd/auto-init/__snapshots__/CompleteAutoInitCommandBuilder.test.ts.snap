// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompleteAutoInitCommandBuilder should create complete auth group given only an auto init config object 1`] = `
Object {
  "customize": Object {},
  "description": "A fake description",
  "examples": Array [
    Object {
      "description": "fake description",
      "options": "fake options",
    },
  ],
  "handler": "fakeHandler",
  "name": "auto-init",
  "options": Array [
    Object {
      "aliases": Array [
        "uc",
      ],
      "defaultValue": false,
      "description": "Save config in the user layer",
      "group": "Automatic Config Initialization Options",
      "name": "user-config",
      "type": "boolean",
    },
    Object {
      "aliases": Array [
        "gc",
      ],
      "defaultValue": false,
      "description": "Save config in the global layer",
      "group": "Automatic Config Initialization Options",
      "name": "global-config",
      "type": "boolean",
    },
    Object {
      "aliases": Array [
        "dr",
        "dry",
      ],
      "defaultValue": false,
      "description": "Display the outcome of the initialization without saving",
      "group": "Automatic Config Initialization Options",
      "name": "dry-run",
      "type": "boolean",
    },
    Object {
      "aliases": Array [
        "e",
      ],
      "defaultValue": false,
      "description": "Open an editor for the configuration",
      "group": "Automatic Config Initialization Options",
      "name": "edit",
      "type": "boolean",
    },
    Object {
      "aliases": Array [
        "ow",
      ],
      "defaultValue": false,
      "description": "Replaces an existing configuration with a new configuration",
      "group": "Automatic Config Initialization Options",
      "implies": Array [
        "for-sure",
      ],
      "name": "overwrite",
      "type": "boolean",
    },
    Object {
      "aliases": Array [
        "f",
      ],
      "defaultValue": false,
      "description": "Confirms the overwrite option",
      "group": "Automatic Config Initialization Options",
      "name": "for-sure",
      "type": "boolean",
    },
    Object {
      "description": "Fake command option",
      "name": "fake",
      "type": "string",
    },
  ],
  "profile": Object {
    "optional": Array [
      "fakeProfile",
    ],
  },
  "summary": "A fake summary",
  "type": "command",
}
`;
